### ISSUE TITLE
Mounting into folders

### ISSUE THREAD

**Issue by jheuking** [openzfsonwindows/ZFSin#97]
ok, learned that you probably can't create mount points on an NTFS drive by using the mount manager 
you need to create a reparse points (as it is implemented now, mentioned [here](https://community.osr.com/discussion/44155/re-how-to-use-ioctl-mountmgr-create-point-without-a-driv-e-letter))
but there are still some issues

1. The folder that was used for mounting needs to be cleaned up on unmount, otherwise remount fails. I tried to delete the folder manually, that makes remount possible. It gets a little complex, when we want to mount other datasets into an already mounted one (pt. 3).
2. There is some code to automatically create a folder, if it doesn't exist, but that doesn't work for several levels. So if you want to mount into C:\mnt\tmp, and neither mnt nor tmp exists, mounting will fail in a bad way. Not sure if this luxury of automatically created folders is provided on other OS, but if we can't/don't want to create the needed folder, we should at least fail gracefully, if it doesn't exist.
3. Mounting inside a mounted folder will fail. If you have a FreeBSD pool (for example), datasets are somewhat interleaved, so it is possible that you have to mount datasets into an already mounted dataset, which is currently not possible. That should produce an error as well as long, as it doesn't work.

**Comment by lundman** [comment]
On other platforms, when mounting libzfs will create the directory tree (hmm does it do tree, or just last dir?) for the mountpoint "/Volumes/pool/mountpoint" (or if it exists, fail if it is not empty).

I tried this at first, but it turns out that on Windows, you can't take a directory, and change it into a reparse-point (it just didn't want to work). So now libzfs (or was it the kernel?) will rmdir the last part "mountpoint" and create a new reparse-point. Admittedly, now I know about SUPERSEDE, you could simply create the reparse-point as that , to both rmdir and create it again.

The kernel code marks it as "delete on close" IIRC, so it should have been deleted automatically, but, I think it would be better to let libzfs handle it. 

In Unix libzfs will also delete the mountpoint directory when unmounting. Unsure why this is not working.

When I last did the mount code, there are three types.  Mounting as driveletter. Mounting as directory inside NTFS, and mounting as directory inside ZFS. All three was working at one point. So if "3" is not working, then we have broken it. The named-mount inside ZFS was a little trickier, since it has to parse the reparse-point data etc.

"4" - this is news to me :)

**Comment by jheuking** [comment]
regarding 3: After experimenting a bit with mountvol, it turns out that you can only create a new mountpoint on a fixed drive, so if you try HDD(USB-STICK(HDD())) it will fail
but you can do HDD(HDD(USB-STICK()))
not sure how the volumes are flagged currently or what special bits are needed to make it a fixed drive

**Comment by jheuking** [comment]
poking around a bit more
if you try to mount into a dataset, it seems to happen like this: first removing the directory fails, then create fails because the directory already exists and in the end, it claims "directory is not a reparse point"
and it seems like the removal fails, because it doesn't search for /mnt but for /MNT in ZwDeleteFile, where it cannot find anything in a case sensitive pool. But we provide '\??\c:\mnt\tmp', so it is probably an issue with ZwDeleteFile.
In Addition to that IO seems not to work so well in a folder, so you cannot delete or create a folder, when the dataset is mounted to C:\mnt for example
Therefore you can't mount C:\mnt\tmp as well, as you can't make delete or create the folder and therefore can't make it a reparse point. And if you try to mount zroot/ROOT/default to a driveletter (e.g. D:), you can't specify that zroot/tmp should be mounted to D:\tmp. 
So in summary, we first need to fix 4. before we can fix 3.

**Comment by jheuking** [comment]
just to keep that in mind here #27

**Comment by tamlin-mike** [comment]
@lundman
> you can't take a directory, and change it into a reparse-point

Provided the directory resides on an NTFS volume you can, but there seems to be some be some arbitrary (and - obviously - undocumented) scenarios when it fails. I think it's if the directory has ever had any contents, even if you delete them, but it's so long since I fought with that my memory is hazy.

Sidenote: Instead of cmd's built-in `mklink`, you could try the `MakeLink` (https://www.codeproject.com/Articles/194/Windows-2000-Junction-Points) I wrote a few decades ago. For some scenarios it just seems to work better - you can f.ex. remove a junction from a directory with it while keeping the directory in place.
